File: src/heap/HeapSort.java
package heap;

import tree.Node;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class HeapSort {
    
    private static BufferedReader reader = null;
    
    // getter
    private static int getInt()
    {
        return Integer.parseInt(getLine());
    }
    
    private static String getLine()
    {
        try {
            return getReader().readLine();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
    
    private static BufferedReader getReader()
    {
        if (reader == null) {
            reader = new BufferedReader(new InputStreamReader(System.in));
        }
        return reader;
    }
    // getter end
    
    public static void main(String[] args)
    {
        int size, i;
        System.out.print("Enter number of items: ");
        size = getInt();
        Heap heap = new Heap(size);
        
        for (i = 0; i < size; i++) {
            int random = ( int ) (Math.random() * 100);
            heap.insertAt(i, new Node.Builder(random, null).build());
            heap.incrementSize();
        }
        
        System.out.print("Random: ");
        heap.displayArray();
        
        for (i = (size / 2) - 1; i >= 0; i--) {
            heap.trickleDown(i);
        }
        
        System.out.print("Heap: ");
        heap.displayArray();
        heap.displayHeap();
        
        for (i = size - 1; i >= 0; i--) {
            Node bigNode = heap.remove();
            heap.insertAt(i, bigNode);
        }
    }
    
}

File: src/heap/HeapSort.kt
package heap

import tree.Node
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

object HeapSort {
    private val reader: BufferedReader by lazy { BufferedReader(InputStreamReader(System.`in`)) }

    @JvmStatic
    fun main(args: Array<String>) {
        print("Enter number of items: ")
        val size = readInt()
        val heap = Heap(size)
        val uniqueNumbers = HashSet<Int>()

        repeat(size) { i ->
            var random: Int
            do {
                random = (Math.random() * 100).toInt() + 1
            } while (!uniqueNumbers.add(random))
            heap.insertAt(i, Node.Builder(random, null).build())
            heap.incrementSize()
        }

        print("Random: ")
        heap.displayArray()

        for (i in (size / 2 - 1) downTo 0) {
            heap.trickleDown(i)
        }

        print("Heap: ")
        heap.displayArray()
        heap.displayHeap()

        for (i in (size - 1) downTo 0) {
            val bigNode = heap.remove()
            heap.insertAt(i, bigNode)
        }
        print("Sorted: ")
        heap.displayArray()
    }

    private fun readInt(): Int = readLineFromUser().toInt()

    private fun readLineFromUser(): String =
        try {
            reader.readLine() ?: throw RuntimeException("No input provided")
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
}
